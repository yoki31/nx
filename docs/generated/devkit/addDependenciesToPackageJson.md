# Function: addDependenciesToPackageJson

â–¸ **addDependenciesToPackageJson**(`tree`, `dependencies`, `devDependencies`, `packageJsonPath?`, `keepExistingVersions?`): [`GeneratorCallback`](../../devkit/documents/GeneratorCallback)

Add Dependencies and Dev Dependencies to package.json

For example:

```typescript
addDependenciesToPackageJson(tree, { react: 'latest' }, { jest: 'latest' });
```

This will **add** `react` and `jest` to the dependencies and devDependencies sections of package.json respectively.

#### Parameters

| Name                    | Type                                  | Description                                                                 |
| :---------------------- | :------------------------------------ | :-------------------------------------------------------------------------- |
| `tree`                  | [`Tree`](../../devkit/documents/Tree) | Tree representing file system to modify                                     |
| `dependencies`          | `Record`\<`string`, `string`\>        | Dependencies to be added to the dependencies section of package.json        |
| `devDependencies`       | `Record`\<`string`, `string`\>        | Dependencies to be added to the devDependencies section of package.json     |
| `packageJsonPath?`      | `string`                              | Path to package.json                                                        |
| `keepExistingVersions?` | `boolean`                             | If true, prevents existing dependencies from being bumped to newer versions |

#### Returns

[`GeneratorCallback`](../../devkit/documents/GeneratorCallback)

Callback to install dependencies only if necessary, no-op otherwise
