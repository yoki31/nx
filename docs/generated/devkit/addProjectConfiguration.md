# Function: addProjectConfiguration

â–¸ **addProjectConfiguration**(`tree`, `projectName`, `projectConfiguration`, `standalone?`): `void`

Adds project configuration to the Nx workspace.

#### Parameters

| Name                   | Type                                                                  | Default value | Description                                                             |
| :--------------------- | :-------------------------------------------------------------------- | :------------ | :---------------------------------------------------------------------- |
| `tree`                 | [`Tree`](../../devkit/documents/Tree)                                 | `undefined`   | the file system tree                                                    |
| `projectName`          | `string`                                                              | `undefined`   | unique name. Often directories are part of the name (e.g., mydir-mylib) |
| `projectConfiguration` | [`ProjectConfiguration`](../../devkit/documents/ProjectConfiguration) | `undefined`   | project configuration                                                   |
| `standalone`           | `boolean`                                                             | `true`        | whether the project is configured in workspace.json or not              |

#### Returns

`void`
