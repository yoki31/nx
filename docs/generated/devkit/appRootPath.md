# Variable: appRootPath

â€¢ `Const` **appRootPath**: `string` = `workspaceRoot`

The root of the workspace.

**`Deprecated`**

use workspaceRoot instead
