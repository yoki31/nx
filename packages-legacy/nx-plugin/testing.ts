export * from '@nx/plugin/testing';
