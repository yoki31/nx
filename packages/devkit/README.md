<p style="text-align: center;">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://raw.githubusercontent.com/nrwl/nx/master/images/nx-dark.svg">
    <img alt="Nx - Smart Monorepos · Fast CI" src="https://raw.githubusercontent.com/nrwl/nx/master/images/nx-light.svg" width="100%">
  </picture>
</p>

{{links}}

<hr>

# Nx: Smart Monorepos · Fast CI

Nx is a build system with built-in tooling and advanced CI capabilities. It helps you maintain and scale monorepos, both locally and on CI.

This package contains a set of utilities for creating Nx plugins.

{{content}}
