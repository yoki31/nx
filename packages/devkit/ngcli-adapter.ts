/**
 * @category Ng CLI Adapter
 */
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
export {
  wrapAngularDevkitSchematic,
  mockSchematicsForTesting,
  NxScopedHost,
} from 'nx/src/adapter/ngcli-adapter';
