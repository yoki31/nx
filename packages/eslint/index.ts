export { lintProjectGenerator } from './src/generators/lint-project/lint-project';
export { lintInitGenerator } from './src/generators/init/init';
export { Linter } from './src/generators/utils/linter';

// @nx/angular needs it for the Angular CLI workspace migration to Nx to
// infer whether a config is using type aware rules and set the
// `hasTypeAwareRules` option of the `@nx/eslint:lint` executor.
export { hasRulesRequiringTypeChecking } from './src/utils/rules-requiring-type-checking';
