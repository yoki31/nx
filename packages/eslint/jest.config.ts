/* eslint-disable */
export default {
  transform: {
    '^.+\\.[tj]sx?$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'html'],
  globals: {},
  displayName: 'eslint',
  preset: '../../jest.preset.js',
};
