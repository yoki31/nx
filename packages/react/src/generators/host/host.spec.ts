import * as devkit from '@nx/devkit';
import type { Tree } from '@nx/devkit';
import { ProjectGraph, readJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import hostGenerator from './host';
import { Linter } from '@nx/eslint';

jest.mock('@nx/devkit', () => {
  const original = jest.requireActual('@nx/devkit');
  return {
    ...original,
    readCachedProjectGraph: jest.fn().mockImplementation(
      (): ProjectGraph => ({
        dependencies: {},
        nodes: {
          test: {
            name: 'test',
            type: 'app',
            data: {
              root: 'test',
              sourceRoot: 'test/src',
              targets: {
                build: {
                  executor: '@nx/webpack:webpack',
                  outputs: ['{options.outputPath}'],
                  defaultConfiguration: 'production',
                  options: {
                    compiler: 'babel',
                    outputPath: 'dist/test',
                    index: 'test/src/index.html',
                    baseHref: '/',
                    main: `test/src/main.tsx`,
                    tsConfig: 'test/tsconfig.app.json',
                    assets: ['test/src/favicon.ico', 'src/assets'],
                    styles: [`test/src/styles.css`],
                    scripts: [],
                    webpackConfig: 'test/webpack.config.js',
                  },
                  configurations: {
                    development: {
                      extractLicenses: false,
                      optimization: false,
                      sourceMap: true,
                      vendorChunk: true,
                    },
                    production: {
                      fileReplacements: [
                        {
                          replace: `test/src/environments/environment.ts`,
                          with: `test/src/environments/environment.prod.ts`,
                        },
                      ],
                      optimization: true,
                      outputHashing: 'all',
                      sourceMap: false,
                      namedChunks: false,
                      extractLicenses: true,
                      vendorChunk: false,
                    },
                  },
                },
                serve: {
                  executor: '@nx/webpack:dev-server',
                  defaultConfiguration: 'development',
                  options: {
                    buildTarget: `test:build`,
                    hmr: true,
                  },
                  configurations: {
                    development: {
                      buildTarget: `test:build:development`,
                    },
                    production: {
                      buildTarget: `test:build:production`,
                      hmr: false,
                    },
                  },
                },
              },
            },
          },
        },
      })
    ),
  };
});

describe('hostGenerator', () => {
  let tree: Tree;

  // TODO(@jaysoo): Turn this back to adding the plugin
  let originalEnv: string;

  beforeEach(() => {
    originalEnv = process.env.NX_ADD_PLUGINS;
    process.env.NX_ADD_PLUGINS = 'false';
  });

  afterEach(() => {
    process.env.NX_ADD_PLUGINS = originalEnv;
  });

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should generate host files and configs when --js=true', async () => {
    await hostGenerator(tree, {
      name: 'test',
      style: 'css',
      linter: Linter.None,
      unitTestRunner: 'none',
      e2eTestRunner: 'none',
      projectNameAndRootFormat: 'as-provided',
      typescriptConfiguration: false,
      skipFormat: true,
      js: true,
    });

    expect(tree.exists('test/tsconfig.json')).toBeTruthy();

    expect(tree.exists('test/src/bootstrap.js')).toBeTruthy();
    expect(tree.exists('test/src/main.js')).toBeTruthy();
    expect(tree.exists('test/src/app/app.js')).toBeTruthy();
  });

  it('should generate host files and configs when --js=false', async () => {
    await hostGenerator(tree, {
      name: 'test',
      style: 'css',
      linter: Linter.None,
      unitTestRunner: 'none',
      e2eTestRunner: 'none',
      projectNameAndRootFormat: 'as-provided',
      typescriptConfiguration: false,
    });

    expect(tree.exists('test/tsconfig.json')).toBeTruthy();

    expect(tree.exists('test/src/bootstrap.tsx')).toBeTruthy();
    expect(tree.exists('test/src/main.ts')).toBeTruthy();
    expect(tree.exists('test/src/app/app.tsx')).toBeTruthy();
  });

  it('should generate host files and configs when --typescriptConfiguration=true', async () => {
    await hostGenerator(tree, {
      name: 'test',
      style: 'css',
      linter: Linter.None,
      unitTestRunner: 'none',
      e2eTestRunner: 'none',
      projectNameAndRootFormat: 'as-provided',
      typescriptConfiguration: true,
      skipFormat: true,
    });

    expect(tree.exists('test/tsconfig.json')).toBeTruthy();

    expect(tree.exists('test/webpack.config.prod.ts')).toBeTruthy();

    expect(tree.exists('test/webpack.config.ts')).toBeTruthy();
    expect(tree.read('test/webpack.config.ts', 'utf-8')).toMatchSnapshot();

    expect(tree.exists('test/module-federation.config.ts')).toBeTruthy();
    expect(
      tree.read('test/module-federation.config.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should generate host files and configs when --typescriptConfiguration=false', async () => {
    await hostGenerator(tree, {
      name: 'test',
      style: 'css',
      linter: Linter.None,
      unitTestRunner: 'none',
      e2eTestRunner: 'none',
      projectNameAndRootFormat: 'as-provided',
      typescriptConfiguration: false,
    });

    expect(tree.exists('test/tsconfig.json')).toBeTruthy();

    expect(tree.exists('test/webpack.config.prod.js')).toBeTruthy();

    expect(tree.exists('test/webpack.config.js')).toBeTruthy();
    expect(tree.read('test/webpack.config.js', 'utf-8')).toMatchSnapshot();

    expect(tree.exists('test/module-federation.config.js')).toBeTruthy();
    expect(
      tree.read('test/module-federation.config.js', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should install @nx/web for the file-server executor', async () => {
    const tree = createTreeWithEmptyWorkspace();
    await hostGenerator(tree, {
      name: 'test',
      style: 'css',
      linter: Linter.None,
      unitTestRunner: 'none',
      e2eTestRunner: 'none',
      projectNameAndRootFormat: 'as-provided',
      skipFormat: true,
    });

    const packageJson = readJson(tree, 'package.json');
    expect(packageJson.devDependencies['@nx/web']).toBeDefined();
  });

  it('should generate host files and configs for SSR', async () => {
    await hostGenerator(tree, {
      name: 'test',
      ssr: true,
      style: 'css',
      linter: Linter.None,
      unitTestRunner: 'none',
      e2eTestRunner: 'none',
      projectNameAndRootFormat: 'as-provided',
      typescriptConfiguration: false,
    });

    expect(tree.exists('test/tsconfig.json')).toBeTruthy();
    expect(tree.exists('test/webpack.config.prod.js')).toBeTruthy();
    expect(tree.exists('test/webpack.server.config.js')).toBeTruthy();
    expect(tree.exists('test/webpack.config.js')).toBeTruthy();
    expect(tree.exists('test/module-federation.config.js')).toBeTruthy();
    expect(tree.exists('test/module-federation.server.config.js')).toBeTruthy();
    expect(tree.exists('test/src/main.server.tsx')).toBeTruthy();
    expect(tree.exists('test/src/bootstrap.tsx')).toBeTruthy();
    expect(tree.exists('test/src/main.ts')).toBeTruthy();

    expect(readJson(tree, 'test/tsconfig.server.json')).toEqual({
      compilerOptions: {
        outDir: '../../out-tsc/server',
        target: 'es2019',
        types: [
          'node',
          '@nx/react/typings/cssmodule.d.ts',
          '@nx/react/typings/image.d.ts',
        ],
      },
      extends: './tsconfig.app.json',
      include: ['src/remotes.d.ts', 'src/main.server.tsx', 'server.ts'],
    });

    expect(
      tree.read('test/webpack.server.config.js', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('test/module-federation.server.config.js', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should generate host files and configs for SSR when --typescriptConfiguration=true', async () => {
    await hostGenerator(tree, {
      name: 'test',
      ssr: true,
      style: 'css',
      linter: Linter.None,
      unitTestRunner: 'none',
      e2eTestRunner: 'none',
      projectNameAndRootFormat: 'as-provided',
      typescriptConfiguration: true,
    });

    expect(tree.exists('test/tsconfig.json')).toBeTruthy();
    expect(tree.exists('test/webpack.config.prod.ts')).toBeTruthy();
    expect(tree.exists('test/webpack.server.config.ts')).toBeTruthy();
    expect(tree.exists('test/webpack.config.ts')).toBeTruthy();
    expect(tree.exists('test/module-federation.config.ts')).toBeTruthy();
    expect(tree.exists('test/module-federation.server.config.ts')).toBeTruthy();
    expect(tree.exists('test/src/main.server.tsx')).toBeTruthy();
    expect(tree.exists('test/src/bootstrap.tsx')).toBeTruthy();
    expect(tree.exists('test/src/main.ts')).toBeTruthy();

    expect(readJson(tree, 'test/tsconfig.server.json')).toEqual({
      compilerOptions: {
        outDir: '../../out-tsc/server',
        target: 'es2019',
        types: [
          'node',
          '@nx/react/typings/cssmodule.d.ts',
          '@nx/react/typings/image.d.ts',
        ],
      },
      extends: './tsconfig.app.json',
      include: ['src/remotes.d.ts', 'src/main.server.tsx', 'server.ts'],
    });

    expect(
      tree.read('test/webpack.server.config.ts', 'utf-8')
    ).toMatchSnapshot();
    expect(
      tree.read('test/module-federation.server.config.ts', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should generate a host and remotes in a directory correctly when using --projectNameAndRootFormat=as-provided', async () => {
    const tree = createTreeWithEmptyWorkspace();

    await hostGenerator(tree, {
      name: 'host-app',
      directory: 'foo/host-app',
      remotes: ['remote1', 'remote2', 'remote3'],
      projectNameAndRootFormat: 'as-provided',
      e2eTestRunner: 'none',
      linter: Linter.None,
      style: 'css',
      unitTestRunner: 'none',
      typescriptConfiguration: false,
    });

    expect(tree.exists('foo/remote1/project.json')).toBeTruthy();
    expect(tree.exists('foo/remote2/project.json')).toBeTruthy();
    expect(tree.exists('foo/remote3/project.json')).toBeTruthy();
    expect(
      tree.read('foo/host-app/module-federation.config.js', 'utf-8')
    ).toContain(`'remote1', 'remote2', 'remote3'`);
  });

  it('should generate a host and remotes in a directory correctly when using --projectNameAndRootFormat=as-provided and --typescriptConfiguration=true', async () => {
    const tree = createTreeWithEmptyWorkspace();

    await hostGenerator(tree, {
      name: 'host-app',
      directory: 'foo/host-app',
      remotes: ['remote1', 'remote2', 'remote3'],
      projectNameAndRootFormat: 'as-provided',
      e2eTestRunner: 'none',
      linter: Linter.None,
      style: 'css',
      unitTestRunner: 'none',
      typescriptConfiguration: true,
    });

    expect(tree.exists('foo/remote1/project.json')).toBeTruthy();
    expect(tree.exists('foo/remote2/project.json')).toBeTruthy();
    expect(tree.exists('foo/remote3/project.json')).toBeTruthy();
    expect(
      tree.read('foo/host-app/module-federation.config.ts', 'utf-8')
    ).toContain(`'remote1', 'remote2', 'remote3'`);
  });

  it('should throw an error if invalid remotes names are provided and --dynamic is set to true', async () => {
    const tree = createTreeWithEmptyWorkspace();
    const remote = 'invalid-remote-name';

    await expect(
      hostGenerator(tree, {
        name: 'myhostapp',
        remotes: [remote],
        dynamic: true,
        projectNameAndRootFormat: 'as-provided',
        e2eTestRunner: 'none',
        linter: Linter.None,
        style: 'css',
        unitTestRunner: 'none',
        typescriptConfiguration: false,
      })
    ).rejects.toThrowError(`Invalid remote name provided: ${remote}.`);
  });
});
