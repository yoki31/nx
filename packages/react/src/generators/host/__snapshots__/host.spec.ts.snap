// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hostGenerator should generate host files and configs for SSR 1`] = `
"const { composePlugins, withNx } = require('@nx/webpack');
const { withReact } = require('@nx/react');
const { withModuleFederationForSSR } = require('@nx/react/module-federation');

const baseConfig = require('./module-federation.config');

const defaultConfig = {
  ...baseConfig,
};

// Nx plugins for webpack to build config object from Nx options and context.
module.exports = composePlugins(
  withNx(),
  withReact({ ssr: true }),
  withModuleFederationForSSR(defaultConfig)
);
"
`;

exports[`hostGenerator should generate host files and configs for SSR 2`] = `
"// @ts-check

/**
 * @type {import('@nx/webpack').ModuleFederationConfig}
 **/
const moduleFederationConfig = {
  name: 'test',
  remotes: [],
};

module.exports = moduleFederationConfig;
"
`;

exports[`hostGenerator should generate host files and configs for SSR when --typescriptConfiguration=true 1`] = `
"import { composePlugins, withNx } from '@nx/webpack';
import { withReact } from '@nx/react';
import { withModuleFederationForSSR } from '@nx/react/module-federation';

import baseConfig from './module-federation.config';

const defaultConfig = {
  ...baseConfig,
};

// Nx plugins for webpack to build config object from Nx options and context.
export default composePlugins(
  withNx(),
  withReact({ ssr: true }),
  withModuleFederationForSSR(defaultConfig)
);
"
`;

exports[`hostGenerator should generate host files and configs for SSR when --typescriptConfiguration=true 2`] = `
"import { ModuleFederationConfig } from '@nx/webpack';

const config: ModuleFederationConfig = {
  name: 'test',
  remotes: [],
};

export default config;
"
`;

exports[`hostGenerator should generate host files and configs when --typescriptConfiguration=false 1`] = `
"const { composePlugins, withNx } = require('@nx/webpack');
const { withReact } = require('@nx/react');
const { withModuleFederation } = require('@nx/react/module-federation');

const baseConfig = require('./module-federation.config');

const config = {
  ...baseConfig,
};

// Nx plugins for webpack to build config object from Nx options and context.
module.exports = composePlugins(
  withNx(),
  withReact(),
  withModuleFederation(config)
);
"
`;

exports[`hostGenerator should generate host files and configs when --typescriptConfiguration=false 2`] = `
"module.exports = {
  name: 'test',
  /**
   * To use a remote that does not exist in your current Nx Workspace
   * You can use the tuple-syntax to define your remote
   *
   * remotes: [['my-external-remote', 'https://nx-angular-remote.netlify.app']]
   *
   * You _may_ need to add a \`remotes.d.ts\` file to your \`src/\` folder declaring the external remote for tsc, with the
   * following content:
   *
   * declare module 'my-external-remote';
   *
   */
  remotes: [],
};
"
`;

exports[`hostGenerator should generate host files and configs when --typescriptConfiguration=true 1`] = `
"import {composePlugins, withNx, ModuleFederationConfig} from '@nx/webpack';
import {withReact} from '@nx/react';
import {withModuleFederation} from '@nx/react/module-federation';

import baseConfig from './module-federation.config';

const config: ModuleFederationConfig = {
    ...baseConfig,
};

// Nx plugins for webpack to build config object from Nx options and context.
export default composePlugins(withNx(), withReact(), withModuleFederation(config));
"
`;

exports[`hostGenerator should generate host files and configs when --typescriptConfiguration=true 2`] = `
"import { ModuleFederationConfig } from '@nx/webpack';

const config: ModuleFederationConfig = {
    name: 'test',
    /**
     * To use a remote that does not exist in your current Nx Workspace
     * You can use the tuple-syntax to define your remote
     *
     * remotes: [['my-external-remote', 'https://nx-angular-remote.netlify.app']]
     *
     * You _may_ need to add a \`remotes.d.ts\` file to your \`src/\` folder declaring the external remote for tsc, with the
     * following content:
     *
     * declare module 'my-external-remote';
     *
     */
    remotes: [
            
  ],
};

export default config;
"
`;
